File: mobile/src/main/java/com/bayapps/android/robophish/ui/PlaybackControlsFragment.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bayapps.android.robophish.ui;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bayapps.android.robophish.MusicService;
import com.bayapps.android.robophish.R;
import com.bayapps.android.robophish.RoboPhishApplicationKt;
import com.squareup.picasso.Picasso;

import javax.inject.Inject;

import timber.log.Timber;

/**
 * A class that shows the Media Queue to the user.
 */
public class PlaybackControlsFragment extends Fragment {
    private ImageButton mPlayPause;
    private TextView mTitle;
    private TextView mSubtitle;
    private TextView mExtraInfo;
    private ImageView mAlbumArt;
    private String mArtUrl;

    @Inject Picasso picasso;

    // Receive callbacks from the MediaController. Here we update our state such as which queue
    // is being shown, the current title and description and the PlaybackState.
    private final MediaControllerCompat.Callback mCallback = new MediaControllerCompat.Callback() {
        @Override
        public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
            Timber.d("Received playback state change to state %s", state.getState());
            PlaybackControlsFragment.this.onPlaybackStateChanged(state);
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            if (metadata == null) {
                return;
            }
            Timber.d("Received metadata state change to mediaId=%s song=%s",
                    metadata.getDescription().getMediaId(),
                    metadata.getDescription().getTitle());
            PlaybackControlsFragment.this.onMetadataChanged(metadata);
        }
    };

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        RoboPhishApplicationKt.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_playback_controls, container, false);

        mPlayPause = rootView.findViewById(R.id.play_pause);
        mPlayPause.setEnabled(true);
        mPlayPause.setOnClickListener(mButtonListener);

        mTitle = rootView.findViewById(R.id.title);
        mSubtitle = rootView.findViewById(R.id.artist);
        mExtraInfo = rootView.findViewById(R.id.extra_info);
        mAlbumArt = rootView.findViewById(R.id.album_art);
        rootView.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), FullScreenPlayerActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
            MediaControllerCompat controller = ((BaseActivity) getActivity())
                    .getSupportMediaController();
            MediaMetadataCompat metadata = controller.getMetadata();
            if (metadata != null) {
                intent.putExtra(MusicPlayerActivity.EXTRA_CURRENT_MEDIA_DESCRIPTION,
                    metadata.getDescription());
            }
            startActivity(intent);
        });
        return rootView;
    }

    @Override
    public void onStart() {
        super.onStart();
        Timber.d("fragment.onStart");
        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        if (controller != null) {
            onConnected();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        Timber.d("fragment.onStop");
        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        if (controller != null) {
            controller.unregisterCallback(mCallback);
        }
    }

    public void onConnected() {
        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        Timber.d("onConnected, mediaController==null? %s", controller == null);
        if (controller != null) {
            onMetadataChanged(controller.getMetadata());
            onPlaybackStateChanged(controller.getPlaybackState());
            controller.registerCallback(mCallback);
        }
    }

    @SuppressLint("BinaryOperationInTimber")
    private void onMetadataChanged(MediaMetadataCompat metadata) {
        Timber.d("onMetadataChanged %s", metadata);
        if (getActivity() == null) {
            Timber.w("onMetadataChanged called when getActivity null," +
                    "this should not happen if the callback was properly unregistered. Ignoring.");
            return;
        }
        if (metadata == null) {
            return;
        }

        mTitle.setText(metadata.getDescription().getTitle());
        mSubtitle.setText(metadata.getDescription().getSubtitle());
        String artUrl = null;
        if (metadata.getDescription().getIconUri() != null) {
            artUrl = metadata.getDescription().getIconUri().toString();
        }
        if (!TextUtils.equals(artUrl, mArtUrl)) {
            mArtUrl = artUrl;
            Bitmap art = metadata.getDescription().getIconBitmap();

            picasso.load(mArtUrl)
                    .fit()
                    .centerInside()
                    .into(mAlbumArt);
        }
    }

    public void setExtraInfo(String extraInfo) {
        if (extraInfo == null) {
            mExtraInfo.setVisibility(View.GONE);
        } else {
            mExtraInfo.setText(extraInfo);
            mExtraInfo.setVisibility(View.VISIBLE);
        }
    }

    @SuppressLint("BinaryOperationInTimber")
    private void onPlaybackStateChanged(PlaybackStateCompat state) {
        Timber.d("onPlaybackStateChanged %s", state);
        if (getActivity() == null) {
            Timber.w("onPlaybackStateChanged called when getActivity null," +
                    "this should not happen if the callback was properly unregistered. Ignoring.");
            return;
        }
        if (state == null) {
            return;
        }
        boolean enablePlay = false;
        switch (state.getState()) {
            case PlaybackStateCompat.STATE_PAUSED:
            case PlaybackStateCompat.STATE_STOPPED:
                enablePlay = true;
                break;
            case PlaybackStateCompat.STATE_ERROR:
                Timber.e("error playbackstate: %s", state.getErrorMessage());
                Toast.makeText(getActivity(), state.getErrorMessage(), Toast.LENGTH_LONG).show();
                break;
        }

        if (enablePlay) {
            mPlayPause.setImageDrawable(
                    ContextCompat.getDrawable(getActivity(), R.drawable.ic_play_arrow_black_36dp));
        } else {
            mPlayPause.setImageDrawable(
                    ContextCompat.getDrawable(getActivity(), R.drawable.ic_pause_black_36dp));
        }

        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        String extraInfo = null;
        if (controller != null && controller.getExtras() != null) {
            String castName = controller.getExtras().getString(MusicService.EXTRA_CONNECTED_CAST);
            if (castName != null) {
                extraInfo = getResources().getString(R.string.casting_to_device, castName);
            }
        }
        setExtraInfo(extraInfo);
    }

    private final View.OnClickListener mButtonListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            MediaControllerCompat controller = ((BaseActivity) getActivity())
                    .getSupportMediaController();
            PlaybackStateCompat stateObj = controller.getPlaybackState();
            final int state = stateObj == null ?
                    PlaybackStateCompat.STATE_NONE : stateObj.getState();
            Timber.d("Button pressed, in state %s", state);
            switch (v.getId()) {
                case R.id.play_pause:
                    Timber.d("Play button pressed, in state %s", state);
                    if (state == PlaybackStateCompat.STATE_PAUSED ||
                            state == PlaybackStateCompat.STATE_STOPPED ||
                            state == PlaybackStateCompat.STATE_NONE) {
                        playMedia();
                    } else if (state == PlaybackStateCompat.STATE_PLAYING ||
                            state == PlaybackStateCompat.STATE_BUFFERING ||
                            state == PlaybackStateCompat.STATE_CONNECTING) {
                        pauseMedia();
                    }
                    break;
            }
        }
    };

    private void playMedia() {
        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        if (controller != null) {
            controller.getTransportControls().play();
        }
    }

    private void pauseMedia() {
        MediaControllerCompat controller = ((BaseActivity) getActivity())
                .getSupportMediaController();
        if (controller != null) {
            controller.getTransportControls().pause();
        }
    }
}

File: mobile/src/main/java/com/bayapps/android/robophish/ui/PlaybackControlsFragment.kt
package com.bayapps.android.robophish.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.bayapps.android.robophish.MusicService
import com.bayapps.android.robophish.R
import com.bayapps.android.robophish.ServiceLocator
import com.squareup.picasso.Picasso
import timber.log.Timber

/**
 * A class that shows the Media Queue to the user.
 */
class PlaybackControlsFragment : Fragment() {
    private lateinit var playPause: ImageButton
    private lateinit var title: TextView
    private lateinit var subtitle: TextView
    private lateinit var extraInfo: TextView
    private lateinit var albumArt: ImageView
    private var artUrl: String? = null

    private val picasso: Picasso by lazy { ServiceLocator.get(requireContext()).picasso }

    private val callback = object : MediaControllerCompat.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat?) {
            Timber.d("Received playback state change to state %s", state?.state)
            onPlaybackStateChangedInternal(state)
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            if (metadata == null) return
            Timber.d(
                "Received metadata state change to mediaId=%s song=%s",
                metadata.description.mediaId,
                metadata.description.title
            )
            onMetadataChangedInternal(metadata)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val rootView = inflater.inflate(R.layout.fragment_playback_controls, container, false)

        playPause = rootView.findViewById(R.id.play_pause)
        playPause.isEnabled = true
        playPause.setOnClickListener(buttonListener)

        title = rootView.findViewById(R.id.title)
        subtitle = rootView.findViewById(R.id.artist)
        extraInfo = rootView.findViewById(R.id.extra_info)
        albumArt = rootView.findViewById(R.id.album_art)
        rootView.setOnClickListener {
            val context = activity ?: return@setOnClickListener
            val intent = Intent(context, FullScreenPlayerActivity::class.java)
                .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
            val controller = (context as BaseActivity).getSupportMediaController()
            val metadata = controller?.metadata
            if (metadata != null) {
                intent.putExtra(
                    MusicPlayerActivity.EXTRA_CURRENT_MEDIA_DESCRIPTION,
                    metadata.description
                )
            }
            startActivity(intent)
        }
        return rootView
    }

    override fun onStart() {
        super.onStart()
        Timber.d("fragment.onStart")
        val controller = (activity as? BaseActivity)?.getSupportMediaController()
        if (controller != null) {
            onConnected()
        }
    }

    override fun onStop() {
        super.onStop()
        Timber.d("fragment.onStop")
        val controller = (activity as? BaseActivity)?.getSupportMediaController()
        controller?.unregisterCallback(callback)
    }

    fun onConnected() {
        val controller = (activity as? BaseActivity)?.getSupportMediaController()
        Timber.d("onConnected, mediaController==null? %s", controller == null)
        if (controller != null) {
            onMetadataChangedInternal(controller.metadata)
            onPlaybackStateChangedInternal(controller.playbackState)
            controller.registerCallback(callback)
        }
    }

    @SuppressLint("BinaryOperationInTimber")
    private fun onMetadataChangedInternal(metadata: MediaMetadataCompat?) {
        Timber.d("onMetadataChanged %s", metadata)
        if (activity == null) {
            Timber.w(
                "onMetadataChanged called when getActivity null, ignoring."
            )
            return
        }
        if (metadata == null) return

        title.text = metadata.description.title
        subtitle.text = metadata.description.subtitle
        val newArtUrl = metadata.description.iconUri?.toString()
        if (!TextUtils.equals(newArtUrl, artUrl)) {
            artUrl = newArtUrl
            val art: Bitmap? = metadata.description.iconBitmap
            if (art != null) {
                albumArt.setImageBitmap(art)
            } else if (!artUrl.isNullOrBlank()) {
                picasso.load(artUrl).fit().centerInside().into(albumArt)
            }
        }
    }

    fun setExtraInfo(extraInfoValue: String?) {
        if (extraInfoValue == null) {
            extraInfo.visibility = View.GONE
        } else {
            extraInfo.text = extraInfoValue
            extraInfo.visibility = View.VISIBLE
        }
    }

    @SuppressLint("BinaryOperationInTimber")
    private fun onPlaybackStateChangedInternal(state: PlaybackStateCompat?) {
        Timber.d("onPlaybackStateChanged %s", state)
        if (activity == null) {
            Timber.w(
                "onPlaybackStateChanged called when getActivity null, ignoring."
            )
            return
        }
        if (state == null) return

        val enablePlay = when (state.state) {
            PlaybackStateCompat.STATE_PAUSED,
            PlaybackStateCompat.STATE_STOPPED -> true
            PlaybackStateCompat.STATE_ERROR -> {
                Timber.e("error playbackstate: %s", state.errorMessage)
                Toast.makeText(activity, state.errorMessage, Toast.LENGTH_LONG).show()
                false
            }
            else -> false
        }

        val drawableId = if (enablePlay) {
            R.drawable.ic_play_arrow_black_36dp
        } else {
            R.drawable.ic_pause_black_36dp
        }
        playPause.setImageDrawable(
            ContextCompat.getDrawable(requireActivity(), drawableId)
        )

        val controller = (activity as? BaseActivity)?.getSupportMediaController()
        val castName = controller?.extras?.getString(MusicService.EXTRA_CONNECTED_CAST)
        val extraInfoValue = if (castName != null) {
            resources.getString(R.string.casting_to_device, castName)
        } else {
            null
        }
        setExtraInfo(extraInfoValue)
    }

    private val buttonListener = View.OnClickListener { view ->
        val controller = (activity as? BaseActivity)?.getSupportMediaController()
        val stateObj = controller?.playbackState
        val state = stateObj?.state ?: PlaybackStateCompat.STATE_NONE
        Timber.d("Button pressed, in state %s", state)
        when (view.id) {
            R.id.play_pause -> {
                Timber.d("Play button pressed, in state %s", state)
                if (state == PlaybackStateCompat.STATE_PAUSED ||
                    state == PlaybackStateCompat.STATE_STOPPED ||
                    state == PlaybackStateCompat.STATE_NONE
                ) {
                    playMedia()
                } else if (state == PlaybackStateCompat.STATE_PLAYING ||
                    state == PlaybackStateCompat.STATE_BUFFERING ||
                    state == PlaybackStateCompat.STATE_CONNECTING
                ) {
                    pauseMedia()
                }
            }
        }
    }

    private fun playMedia() {
        (activity as? BaseActivity)?.getSupportMediaController()
            ?.transportControls
            ?.play()
    }

    private fun pauseMedia() {
        (activity as? BaseActivity)?.getSupportMediaController()
            ?.transportControls
            ?.pause()
    }
}
